### PR TITLE
tests: Try to fix installed tests

### PR DESCRIPTION
**Pull request by jtojnar**
Partial fix for #502

### CODE REVIEW

File: tests/test-document-fuse.sh [hunk 1]
```diff
@@ -49,20 +49,24 @@ export XDG_RUNTIME_DIR=${TEST_DATA_DIR}/runtime
...
     rm -rf $TEST_DATA_DIR
 }
 trap cleanup EXIT
 
-sed s#@testdir@#${test_builddir}# ${test_srcdir}/session.conf.in > session.conf
+conf_template="${test_srcdir}/session.conf.in"
+if [[ ! -f "$conf_template" ]]; then
+    conf_template="${test_srcdir}/tests/session.conf.in"
+fi
+sed s#@testdir@#${test_builddir}# $conf_template > session.conf
```
Comment (jtojnar) [inline tests/test-document-fuse.sh:62]: This is quite ugly. It would be better if we could make install to the same directory but I do not understand why

https://github.com/flatpak/xdg-desktop-portal/blob/b907d57fd6e45bebb4aca71f50b7017dbea7fc5a/tests/Makefile.am.inc#L112-L113

places it into `tests` subdirectory, while

https://github.com/flatpak/xdg-desktop-portal/blob/b907d57fd6e45bebb4aca71f50b7017dbea7fc5a/tests/Makefile.am.inc#L116-L117

does not.

Comment (jtojnar) [reply tests/test-document-fuse.sh:62]: Reading `glib-tap.mk`:

> Note that _data is always used with the nobase_ automake variable name to ensure that installed test data is installed in the same way as it appears in the package layout.

Is adding it to `installed_test_DATA` and `EXTRA_DIST` instead an acceptable solution?

**Comment by jtojnar** [comment]
Given how often the installed tests are broken, we should probably add them to CI too.

File: tests/Makefile.am.inc [hunk 2]
```diff
@@ -109,6 +109,9 @@ dist_installed_test_dbs_DATA = tests/dbs/no_tables
 dist_installed_test_portals_DATA = tests/portals/test.portal
 endif
 
+tests/test-document-fuse.sh: tests/test-document-fuse.sh.in
+	$(AM_V_GEN) $(SED) -e "s|\@libexecdir\@|$(libexecdir)|" $< > $@
```
Comment (smcv) [inline tests/Makefile.am.inc:113]: It seems confusing that this is replacing `@libexecdir@` as an Autoconf-style substitution, while leaving `@testdir@` as a literal string.

File: tests/test-document-fuse.sh.in [hunk 3]
```diff
@@ -72,7 +72,12 @@ fi
 
 # Run portal manually so that we get any segfault our assert output
 # Add -v here to get debug output from fuse
-./xdg-document-portal -r &
+if [ -n "${XDP_UNINSTALLED:-}" ]; then
+    document_portal=./xdg-document-portal
+else
+    document_portal=@libexecdir@/xdg-document-portal
+fi
+"${document_portal}" -r &
```
Comment (smcv) [inline tests/test-document-fuse.sh.in:80]: If the document portal is properly D-Bus-activatable (which it should be, when running installed-tests, because being D-Bus-activatable is part of the interface it is intended to provide) then running the document portal "by hand" like this is actually unnecessary, because it will be D-Bus-activated anyway. Running it in the background means that if it fails to start, the error is ignored anyway.

Perhaps this would be better:

```
if [ -n "${XDP_UNINSTALLED:-}" ]; then
    "${document_portal}" -r &
fi
# else start it via D-Bus activation as usual
```

That would have two advantages:

* if it isn't D-Bus-activatable, for example because upstream changes broke the `.service` file or a downstream forgot to package it, then the test will correctly fail
* the `@libexecdir@` substitution is no longer necessary and the shell script can go back to being directly runnable

File: tests/test-document-fuse.sh [hunk 4]
```diff
@@ -49,7 +49,7 @@ export XDG_RUNTIME_DIR=${TEST_DATA_DIR}/runtime
 cleanup () {
     fusermount -u $XDG_RUNTIME_DIR/doc || :
     sleep 0.1
-    /bin/kill -9 $DBUS_SESSION_BUS_PID
+    kill -9 $DBUS_SESSION_BUS_PID
```
Comment (smcv) [inline tests/test-document-fuse.sh:52]: This change seems fine. There are some corner-case differences between coreutils kill and bash's kill builtin, but they don't seem like they would be applicable here.

As a side suggestion, killing the dbus-daemon with the default `SIGTERM` will let it shut down gracefully (which matters in some rare situations involving non-abstract AF_UNIX sockets), and the variable is underquoted:

```suggestion
    kill "$DBUS_SESSION_BUS_PID"
```

File: tests/test-document-fuse.sh (hunk 1 above)
Comment (smcv) [reply tests/test-document-fuse.sh:62]: I think this is because `glib-tap.mk` was designed with recursive `make` in mind, but xdg-desktop-portal is using non-recursive `make`, so the expectations are different. Further complicating things, the "upstream" version of `glib-tap.mk` was removed from GLib 2.59 with the removal of Autotools, and the version in xdg-desktop-portal has been patched anyway.

> Is adding it to `installed_test_DATA` and `EXTRA_DIST` instead an acceptable solution?

I would personally put it in `dist_installed_test_DATA`, with the capitals, instead (which is more or less equivalent to the combination of `installed_test_DATA` and `EXTRA_DIST`, but without the duplication).

File: tests/test-document-fuse.sh.in [hunk 5]
```diff
@@ -49,37 +49,46 @@ export XDG_RUNTIME_DIR=${TEST_DATA_DIR}/runtime
...
 
 # First run a basic single-thread test
 echo Testing single-threaded
-python3 ${test_srcdir}/test-document-fuse.py --iterations 3 -v
+${test_srcdir}/test-document-fuse.py --iterations 3 -v
```
Comment (smcv) [inline tests/test-document-fuse.sh.in:84]: This will break build-time tests, unless you chmod it +x.

File: tests/test-document-fuse.sh.in (hunk 3 above)
Comment (jtojnar) [reply tests/test-document-fuse.sh.in:80]: Does the comment "Run portal manually so that we get any segfault our assert output" not apply then?

File: tests/Makefile.am.inc (hunk 2 above)
Comment (jtojnar) [reply tests/Makefile.am.inc:113]: Indeed. Should I add a comment or escape it as `@@testdir@@`?

File: tests/test-document-fuse.sh.in (hunk 3 above)
Comment (smcv) [reply tests/test-document-fuse.sh.in:80]: If I was debugging in the sort of detail that requires that, I'd expect to be using the build-time tests.

For installed-tests, I think it's reasonable to expect that debugging failures will require whole-system debugging tools, for example capturing core dumps with systemd's `systemd-coredumpd` and `coredumpctl`.

**Comment by smcv** [comment]
I should say here that I'm not a maintainer, and actual maintainers might overrule me on any of this.

I noticed while updating the Debian package that we have been skipping the FUSE test because the minimal test VM doesn't have `capsh`, so I wouldn't have seen several of the failure modes you're seeing. I'm testing patches that do what I suggested above.

**Comment by smcv** [comment]
> Given how often the installed tests are broken, we should probably add them to CI too.

That might not be possible for the document-portal test, depending on how Github runs tests - FUSE is usually not available in containers. I agree the other tests should be run in CI though.

**Comment by smcv** [comment]
Running build-time tests has been failing with timeouts in CI since May, so I don't think running the installed-tests would actually do any better.

File: tests/Makefile.am.inc (hunk 2 above)
Comment (smcv) [reply tests/Makefile.am.inc:113]: I'd prefer to see whether we can make it unnecessary first.

**Comment by smcv** [comment]
#521 is another try at this, taking some commits directly from this PR, applying my suggestions, and adding some more CI-related fixes.

**Comment by smcv** [comment]
Now that #521 has been merged, I think this MR is no longer necessary.

**Comment by jtojnar** [comment]
Thanks for your work.
